filters/crop/Point.cpp

```diff
@@ -0,0 +1,284 @@
+{
+    while (p((char)in.get()))
+        ;
+    if (in.eof())
+        in.clear(in.rdstate() & ~std::ios::failbit);
+    else
+        in.unget();
+}
+
+bool eat(std::istream& in, char c)
+{
+    if ((char)in.get() == c)
+        return true;
+    in.unget();
+    return false;
+}
+
+void readxy(std::istream& istr, double& x, double& y)
+{
+    eat(istr, isspace);
+    if (!eat(istr,'('))
+        istr.setstate(std::ios_base::failbit);
+
+    eat(istr, isspace);
+    istr >> x;
+
+    eat(istr, isspace);
+    if (!eat(istr,','))
+        istr.setstate(std::ios_base::failbit);
+
+    eat(istr, isspace);
+    istr >> y;
+
+    eat(istr, isspace);
+    if (!eat(istr,')'))
+        istr.setstate(std::ios_base::failbit);
+}
+
+void readxyz(std::istream& istr, double& x, double& y, double& z)
+{
+    eat(istr, isspace);
+    if (!eat(istr,'('))
+        istr.setstate(std::ios_base::failbit);
+
+    eat(istr, isspace);
+    istr >> x;
+
+    eat(istr, isspace);
+    if (!eat(istr,','))
+        istr.setstate(std::ios_base::failbit);
+
+    eat(istr, isspace);
+    istr >> y;
+
+    eat(istr, isspace);
+    if (!eat(istr,','))
+        istr.setstate(std::ios_base::failbit);
+
+    eat(istr, isspace);
+    istr >> z;
+
+    eat(istr, isspace);
+    if (!eat(istr,')'))
+        istr.setstate(std::ios_base::failbit);
+}
+
+
+
+namespace cropfilter
+{
+
+bool Point2D::empty() const
+{
+    return  x == LOWEST && y == LOWEST;
+}
+
+void Point2D::clear()
+{
+    x = LOWEST; y = LOWEST;
+}
+
+void Point3D::clear()
+{
+    Point2D::clear();
+    z = LOWEST;
+}
+
+bool Point3D::empty() const
+{
+    return  x == LOWEST && y == LOWEST && z == LOWEST;
+}
+
+Point::Point(const Point3D& point) : m_point(point)
+{}
+
+
+Point::Point(const Point2D& point) : m_point(point)
+{
+    m_point.z = LOWEST;
+}
+
+bool Point::is3d() const
+{
+    return (m_point.z != LOWEST );
+}
+
+
+void Point::set(const Point3D& point)
+{
+    m_point = point;
+}
+
+
+void Point::set(const Point2D& point)
+{
+    m_point = Point3D(point);
+    m_point.z = LOWEST;
+}
+
+Point2D Point::to2d() const
+{
+    return m_point.to2d();
+}
+
+Point2D Point3D::to2d() const
+{
+    return (*this);
+}
+
+Point3D Point::to3d() const
+{
+    if (m_point.x == LOWEST && m_point.y == LOWEST)
+        return Point3D();
+    return m_point;
+}
+
+
+std::istream& operator>>(std::istream& istr, cropfilter::Point2D& point)
+{
+    char left_paren = (char)istr.get();
+    if (!istr.good())
+    {
+        istr.setstate(std::ios_base::failbit);
+        return istr;
+    }
+    const char right_paren = (char)istr.get();
+
+    if (left_paren == '(' && right_paren == ')')
+    {
+        point = cropfilter::Point2D();
+        return istr;
+    }
+    istr.unget();
+    istr.unget(); // ()
+
+    double x, y;
+
+    readxy(istr, x, y);
+
+    if (istr.good())
+    {
+        point.x = x;
+        point.y = y;
+    }
+    return istr;
+}
+
+std::istream& operator>>(std::istream& istr, cropfilter::Point3D& point)
```

Comment:
I think it would be nice for user if you could tell them what's wrong if you detect an error as you're parsing and you know the problem.